### PR TITLE
New functions: Set-CIisConfigurationAttribute, Remove-CIisConfigurationAttribute, ConvertTo-CIisVirtualPath

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

File: Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1 [hunk 1]
```diff
@@ -29,53 +29,38 @@ function Disable-CIisSecurityAuthentication
...
     #>
-    [CmdletBinding(SupportsShouldProcess=$true)]
+    [CmdletBinding(SupportsShouldProcess)]
     param(
-        [Parameter(Mandatory=$true)]
-        [string]
-        # The site where anonymous authentication should be set.
-        $SiteName,
-        
+        # The site where authentication should be set.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1]: Replace `set` with `disabled`

File: Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1 [hunk 2]
```diff
@@ -29,53 +29,38 @@ function Disable-CIisSecurityAuthentication
...
+        [Parameter(Mandatory)]
+        [String] $SiteName,
+
+        # The optional path where authentication should be set.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1]: Replace `set` with `disabled`

File: Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1 [hunk 3]
```diff
@@ -29,53 +29,38 @@ function Disable-CIisSecurityAuthentication
...
-        [Parameter(Mandatory=$true,ParameterSetName='Anonymous')]
-        [Switch]
-        # Enable anonymouse authentication.
-        $Anonymous,
-        
-        [Parameter(Mandatory=$true,ParameterSetName='Basic')]
-        [Switch]
+        [String] $VirtualPath = '',
+
+        # Enable anonymous authentication.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1]: Replace `Enable` with `Disable`

File: Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1 [hunk 4]
```diff
@@ -29,53 +29,38 @@ function Disable-CIisSecurityAuthentication
...
+        [Parameter(Mandatory, ParameterSetName='anonymousAuthentication')]
+        [switch] $Anonymous,
+
         # Enable basic authentication.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1]: Replace `Enable` with `Disable`

File: Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1 [hunk 5]
```diff
@@ -29,53 +29,38 @@ function Disable-CIisSecurityAuthentication
...
-        $Basic,
-        
-        [Parameter(Mandatory=$true,ParameterSetName='Windows')]
-        [Switch]
+        [Parameter(Mandatory, ParameterSetName='basicAuthentication')]
+        [switch] $Basic,
+
         # Enable Windows authentication.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Disable-CIisSecurityAuthentication.ps1]: Replace `Enable` with `Disable`

File: Carbon.IIS/Functions/Join-CIisVirtualPath.ps1 [hunk 6]
```diff
@@ -3,38 +3,50 @@ function Join-CIisVirtualPath
...
     .DESCRIPTION
-    Removes extra slashes.  Converts backward slashes to forward slashes.  Relative portions are not removed.  Sorry.
-
-    Beginning with Carbon 2.0.1, this function is available only if IIS is installed.
+    Removes extra slashes and relative path signifiers (e.g. `.` and `..`).  Converts backward slashes to forward
+    slashes.
 
     .EXAMPLE
-    Join-CIisVirtualPath 'SiteName' 'Virtual/Path'
+    Join-CIisVirtualPath -Path 'SiteName' -Join 'Virtual/Path'
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Join-CIisVirtualPath.ps1]: `-Join` should be `-ChildPath`

File: Carbon.IIS/Functions/Join-CIisVirtualPath.ps1 [hunk 7]
```diff
@@ -3,38 +3,50 @@ function Join-CIisVirtualPath
...
-        [string]
         # The parent path.
-        $Path,
+        [Parameter(Mandatory, Position=0)]
+        [String]$Path,
 
-        [Parameter(Position=1)]
-        [string]
-        $ChildPath
+        [Parameter(Mandatory, Position=1)]
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Join-CIisVirtualPath.ps1:24]: Missing parameter doc

File: Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1 [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    attribute to remove/delete to the `Name` parameter. The function deletes that attribute. If the attribute doesn't
+    exist, nothing happens.
+
+    To delete more than one attribute on a specific element at a ttime, either pass multiple names to the `Name`
+    parameter, or pipe the list of attributes to `Remove-CIisConfigurationAttribute`.
+
+    To delete/remove an attribute from the configuration of an application/virtual directory under a website, pass the
+    application/virtual diretory's name/path to the `VirtualPath` parameter.
+
+    The current value of the attribute is written to the information stream. To mask the attribute's value in output,
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1:21]: Could add a note that if the attribute is `Password` it is always masked.

File: Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1 [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            $msgPrefix = "$($basePrefix)@$($nameFormat -f $nameItem)]  "
+
+            Write-Debug "$($msgPrefix)$($attr.IsInheritedFromDefaultValue)  $($attr.Value)  $($attr.Schema.DefaultValue)"
+            $hasDefaultValue = $attr.Value -eq $attr.Schema.DefaultValue
+            if( -not $attr.IsInheritedFromDefaultValue -and -not $hasDefaultValue )
+            {
+                $currentValue = $attr.Value
+                if( $Sensitive -or $nameItem -eq 'Password' )
+                {
+                    $currentValue = '*' * $currentValue.ToString().Length
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1]: Should we mask with a static length regardless of original value length for extra security?

File: Carbon.IIS/Functions/Set-CIisAnonymousAuthentication.ps1 [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        # `$false`, e.g. `-Enabled:$false`.
+        [switch] $Enabled,
+
+        # The username of the identity to use to run anonymous requests.
+        [String] $UserName,
+
+        # The password username of the identity to use to run anonymous requests. Not needed if using system accounts.
+        [SecureString] $Password,
+
+        # The logon method to use fo anonymous access.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Set-CIisAnonymousAuthentication.ps1]: Typo: `fo` should be `for`

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+    .EXAMPLE
+    Set-CIisConfigurationAttribute -SiteName 'SiteOne' -VirtualPath 'old_app' -SectionPath 'system.webServer/httpRedirect' -Name 'destination' -Value 'http://example.com'
+
+    Demonstrates how to set attribute values on a sub-path in a website by passing the path to the `VirtualPath`
+    parameter.
+    #>
+    [CmdletBinding(SupportsShouldProcess)]
+    param(
+        # The name of the website whose anonymous authentication settings to configure.
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Should be `Name of the website whose attribute values to configure.`

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        [Parameter(Mandatory)]
+        [String] $SiteName,
+
+        # The virtual path to a directory, application, or virtual directory whose anonymous authentication settings
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Should be `whose attribute values to configure.`

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+
+        if( $Value -is [Enum] )
+        {
+            $currentValueMsg = [enum]::GetName($Value.GetType().FullName, $currentValue)
+        }
+
+        if( $protectValue )
+        {
+            $currentValueMsg = '*' * $currentValueMsg.ToString().Length
+            $valueMsg = '*' * $valueMsg.ToString().Length
```
Comment (decoyjoe) [inline Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Always mask with a static length for extra security?

**Review by decoyjoe — APPROVED** [review]

File: Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1 (hunk 9 above)
Comment (splatteredbits) [reply Carbon.IIS/Functions/Remove-CIisConfigurationAttribute.ps1]: Yeah. Probably.

File: Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1 (hunk 13 above)
Comment (splatteredbits) [reply Carbon.IIS/Functions/Set-CIisConfigurationAttribute.ps1]: Yes.

**Pull request merged**